File: 2024/ruby/day01/day01.rb
module Day01
  def part1(input)
    left_list, right_list = self.parse(input)
    left_list = left_list.sort
    right_list = right_list.sort
    return left_list.zip(right_list).map { |left, right| (left - right).abs }.sum.to_s
  end

  def part2(input)
    left_list, right_list = parse(input)
    frequency = Hash.new(0)
    right_list.each { |n| frequency[n] += 1 }
    return left_list.map { |n| n * frequency[n] }.sum.to_s
  end

  private
  
  def parse(input)
    left_list = []
    right_list = []
    for line in input.lines do
      parts = line.split("   ")
      left_list.push(parts[0].to_i)
      right_list.push(parts[1].to_i)
    end

    return left_list, right_list
  end

  module_function :part1, :part2, :parse
end

File: 2024/elixir/lib/day01/day01.ex
defmodule Day01 do
  def part1(input) do
    {left, right} = parse(input)

    left = left |> Enum.sort()
    right = right |> Enum.sort()

    Enum.zip(left, right)
    |> Enum.map(fn {l, r} -> l - r end)
    |> Enum.map(&abs/1)
    |> Enum.sum()
    |> Integer.to_string()
  end

  def part2(input) do
    {left, right} = parse(input)
    freq = Enum.frequencies(right)
    left |> Enum.map(fn n -> n * Map.get(freq, n, 0) end) |> Enum.sum() |> Integer.to_string()
  end

  def parse(input) do
    entries =
      input
      |> String.split("\n", trim: true)
      |> Enum.map(fn line ->
        line |> String.split(" ", trim: true) |> Enum.map(&String.to_integer/1)
      end)

    left = entries |> Enum.map(fn [x, _] -> x end)
    right = entries |> Enum.map(fn [_, x] -> x end)
    {left, right}
  end
end
